src/Microsoft.AspNet.OData.Shared/DeltaSetOfT.cs

```diff
@@ -0,0 +1,357 @@
+            return deltaSet;
+        }
+
+        private IDeltaSetItem HandleFailedOperation(Delta<TStructuralType> changedObj, DataModificationOperationKind operation, TStructuralType originalObj, string errorMessage)
+        {
+            IDeltaSetItem deltaSetItem = null;
+            DataModificationExceptionType dataModificationExceptionType = new DataModificationExceptionType(operation);
+            dataModificationExceptionType.MessageType = new MessageType();
+            dataModificationExceptionType.MessageType.Message = errorMessage;
+
+            // This handles the Data Modification exception. This adds Core.DataModificationException annotation and also copy other instance annotations.
+            //The failed operation will be based on the protocol
+            switch (operation)
+            {
+                case DataModificationOperationKind.Update:
+                    deltaSetItem = changedObj;
+                    break;
+                case DataModificationOperationKind.Insert:
+                    {
+                        deltaSetItem = CreateDeletedEntityForFailedOperation(changedObj);
+
+                        break;
+                    }
+                case DataModificationOperationKind.Delete:
+                    {
+                        deltaSetItem = CreateEntityObjectforFailedOperation(changedObj, originalObj);                        
+                        break;
+                    }
+            }
+
+
+            deltaSetItem.TransientInstanceAnnotationContainer = changedObj.TransientInstanceAnnotationContainer;
+            deltaSetItem.TransientInstanceAnnotationContainer.AddResourceAnnotation("Core.DataModificationException", dataModificationExceptionType);
+
+            Contract.Assert(deltaSetItem != null);
+
+            return deltaSetItem;
+        }
+
+        private IDeltaSetItem CreateEntityObjectforFailedOperation(Delta<TStructuralType> changedObj, TStructuralType originalObj)
+        {
+            Type type = typeof(Delta<>).MakeGenericType(_clrType);
+
+            Delta<TStructuralType> deltaObject = Activator.CreateInstance(type, _clrType, _clrType.GetProperties().Select(x=>x.Name), null,
+                changedObj.InstanceAnnotationsPropertyInfo) as Delta<TStructuralType>;
+
+            SetProperties(originalObj, deltaObject);
+
+            if (deltaObject.InstanceAnnotationsPropertyInfo != null) {
+
+                object instAnnValue;
+                changedObj.TryGetPropertyValue(deltaObject.InstanceAnnotationsPropertyInfo.Name, out instAnnValue);
+                IODataInstanceAnnotationContainer instanceAnnotations = instAnnValue as IODataInstanceAnnotationContainer;
+
+                if (instanceAnnotations != null)
+                {
+                    deltaObject.TrySetPropertyValue(deltaObject.InstanceAnnotationsPropertyInfo.Name, instanceAnnotations);
```

Comment:
>TrySetPropertyValue [](start = 32, length = 19)

This assumes the instance annotation property is settable, rather than copying to the deltaObject's existing IODataInstanceAnnotationContainer.

Reply:
Yes, but thats how we set the instance annotations on Delta<T>, since we are actually setting hte property of T, right and the container for transient annotations are not exposed